File: clients/java/client-impl/src/main/java/com/flipkart/grayskull/utils/RetryUtil.java
package com.flipkart.grayskull.utils;

import com.flipkart.grayskull.constants.MDCKeys;
import com.flipkart.grayskull.models.exceptions.GrayskullException;
import com.flipkart.grayskull.models.exceptions.RetryableException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;

/**
 * Utility class for retrying operations with exponential backoff.
 * <p>
 * This class implements a retry mechanism with exponential backoff for operations
 * that may fail transiently. It will retry operations that throw {@link RetryableException}
 * up to a maximum number of attempts, with increasing delays between attempts.
 * The wait time is capped at a maximum of 1 minute.
 * </p>
 */
public final class RetryUtil {
    private static final Logger log = LoggerFactory.getLogger(RetryUtil.class);
    private static final long MAX_WAIT_TIME_MS = 60000; // 1 minute

    private final int maxAttempt;
    private final int interval;

    /**
     * Constructor to initialize the RetryUtil with the number of attempts and the interval between retries.
     *
     * @param maxAttempt The maximum number of retry attempts.
     * @param interval   The initial interval between retries in milliseconds.
     */
    public RetryUtil(int maxAttempt, int interval) {
        this.maxAttempt = maxAttempt;
        this.interval = interval;
    }

    /**
     * Executes the given task with retry logic.
     * <p>
     * If the task throws a {@link RetryableException}, it will be retried up to maxAttempt times
     * with exponential backoff. If all retry attempts are exhausted, the {@link RetryableException}
     * is wrapped in a {@link GrayskullException}. Non-retryable exceptions are thrown immediately.
     * The wait time between retries is capped at a maximum of 1 minute.
     * </p>
     *
     * @param task The task to execute
     * @param <T>  The return type of the task
     * @return The result of the task
     * @throws GrayskullException if the task fails after all retry attempts
     * @throws Exception if the task throws a non-retryable exception
     */
    public <T> T retry(CheckedSupplier<T> task) throws Exception {
        long currentInterval = interval;
        Exception lastException = null;
        String requestId = MDC.get(MDCKeys.GRAYSKULL_REQUEST_ID);
        
        for (int attempt = 1; attempt <= maxAttempt; attempt++) {
            try {
                log.debug("[RequestId:{}] Executing task, attempt {} of {}", requestId, attempt, maxAttempt);
                T result = task.get();
                if (attempt > 1) {
                    log.info("[RequestId:{}] Task succeeded on attempt {}", requestId, attempt);
                }
                return result;
            } catch (RetryableException e) {
                lastException = e;
                log.warn("[RequestId:{}] Retryable exception on attempt {} of {}: {}", requestId, attempt, maxAttempt, e.getMessage());
                
                if (attempt == maxAttempt) {
                    log.error("[RequestId:{}] Max retry attempts reached ({}), throwing exception", requestId, maxAttempt);
                    throw new GrayskullException(e.getStatusCode(), "Failed after " + maxAttempt + " retry attempts: " + e.getMessage(), e);
                }
                
                // Sleep before next retry (exponential backoff)
                try {
                    log.debug("[RequestId:{}] Waiting {}ms before retry", requestId, currentInterval);
                    Thread.sleep(currentInterval);
                } catch (InterruptedException ie) {
                    Thread.currentThread().interrupt();
                    throw new GrayskullException("Retry interrupted", ie);
                }
                currentInterval = Math.min(currentInterval * 2, MAX_WAIT_TIME_MS); // Exponential backoff capped at 1 minute
            } 
        }
        throw new GrayskullException("Reached maximum attempts: " + maxAttempt, lastException);
    }

    /**
     * Functional interface representing a supplier that can throw an exception.
     *
     * @param <T> The type of the result supplied by this supplier.
     */
    @FunctionalInterface
    public interface CheckedSupplier<T> {
        T get() throws Exception;
    }
}

File: clients/go/client_impl/utils/retryUtil.go
package utils

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"math"
	"time"

	"github.com/grayskull/client_impl/constants"
	"github.com/grayskull/client_impl/models/exceptions"
)

const (
	maxWaitTime = 1 * time.Minute // 1 minute
)

// RetryUtil implements a retry mechanism with exponential backoff.
// It retries operations that return RetryableError up to a maximum number of attempts,
// with increasing delays between attempts. The wait time is capped at 1 minute.
type RetryUtil struct {
	maxAttempts int
	initialWait time.Duration
	logger      *slog.Logger
}

// NewRetryUtil creates a new RetryUtil with the specified parameters.
func NewRetryUtil(maxAttempts int, initialWait time.Duration, logger *slog.Logger) *RetryUtil {
	if logger == nil {
		logger = slog.Default()
	}
	return &RetryUtil{
		maxAttempts: maxAttempts,
		initialWait: initialWait,
		logger:      logger,
	}
}

// Retry executes the provided function with retry logic.
// It will retry on RetryableError up to the maximum number of attempts.
func (r *RetryUtil) Retry(ctx context.Context, fn func() (interface{}, error)) (interface{}, error) {
	currentWait := r.initialWait

	for attempt := 1; attempt <= r.maxAttempts; attempt++ {
		// Get request ID from context
		requestID := ""
		if ctx != nil {
			if v := ctx.Value(constants.GrayskullRequestID); v != nil {
				if id, ok := v.(string); ok {
					requestID = id
				}
			}
		}

		// Log attempt
		r.logger.DebugContext(ctx, "executing task",
			"attempt", fmt.Sprintf("%d/%d", attempt, r.maxAttempts),
			"requestID", requestID,
		)

		// Execute the function
		result, err := fn()
		if err == nil {
			if attempt > 1 {
				r.logger.InfoContext(ctx, "task succeeded on retry",
					"attempt", attempt,
					"requestID", requestID,
				)
			}
			return result, nil
		}

		// Check if error is retryable
		var retryableErr *exceptions.RetryableError
		if !errors.As(err, &retryableErr) {
			return nil, err // Return non-retryable errors immediately
		}

		if attempt == r.maxAttempts {
			r.logger.ErrorContext(ctx, "max retry attempts reached",
				"maxAttempts", r.maxAttempts,
				"requestID", requestID,
				"error", err,
			)
			return nil, exceptions.NewGrayskullErrorWithCause(
				retryableErr.StatusCode,
				fmt.Sprintf("failed after %d retry attempts", r.maxAttempts),
				err,
			)
		}

		// Log warning and wait before retry
		r.logger.WarnContext(ctx, "retryable error, will retry",
			"attempt", fmt.Sprintf("%d/%d", attempt, r.maxAttempts),
			"waitTime", currentWait,
			"requestID", requestID,
			"error", err,
		)

		// Wait with exponential backoff
		select {
		case <-time.After(currentWait):
			// Continue with next attempt
		case <-ctx.Done():
			return nil, exceptions.NewGrayskullErrorWithCause(0, "retry operation canceled", ctx.Err())
		}

		// Calculate next wait time with exponential backoff, capped at maxWaitTime
		currentWait = time.Duration(math.Min(float64(currentWait*2), float64(maxWaitTime)))
	}
	// Return an error when max attempts are reached without success
	return nil, fmt.Errorf("maximum number of retry attempts (%d) reached", r.maxAttempts)
}
